<!DOCTYPE html>
<html xmlns:th="http://www.thymeleaf.org">
<head>
    <meta charset="UTF-8"></meta>
    <title th:inline="text">[[#{app.name}]] :: Help</title>

    <link th:href="@{/webjars/bootstrap/3.3.6/css/bootstrap.min.css}" rel="stylesheet" media="screen"/>
    <link rel="stylesheet" th:href="@{/css/application.css}"/>
    <link rel="icon" th:href="@{/images/oops.ico}"/>

    <script th:src="@{/webjars/jquery/2.1.3/jquery.min.js}"></script>
    <script th:src="@{/js/jquery.ui.widget.js}"></script>
    <script th:src="@{/js/jquery.iframe-transport.js}"></script>
    <script th:src="@{/js/jquery.fileupload.js}"></script>
    <script th:src="@{/js/bootstrap.min.js}"></script>
    <script th:src="@{/js/script.js}"></script>

</head>

<body>
<div th:replace="fragments/header :: header"></div>

<div class="container">
    <div class="div-ctr">
        <h2 th:inline="text">antiSMASH - OOPS HELP</h2>
    </div>
    <br/>

    <div style="display: table-cell; padding-right: 30px; border-right: solid; border-color: lightgrey">
        <ul class="nav nav-pills nav-stacked">
            <li class="active">
                <a data-toggle="pill" href="#helpHome">antiSMASH - OOPS</a>
            </li>
            <li>
                <a data-toggle="pill" href="#nog">Number of Genes</a>
            </li>
            <li>
                <a data-toggle="pill" href="#sl">CDS Length</a>
            </li>
            <li>
                <a data-toggle="pill" href="#gcc">GC Content</a>
            </li>
            <li>
                <a data-toggle="pill" href="#cb">Codon Bias</a>
            </li>
            <li>
                <a data-toggle="pill" href="#rs">Reference Species</a>
            </li>
            <li>
                <a data-toggle="pill" href="#ct">Cluster Type</a>
            </li>
            <li>
                <a data-toggle="pill" href="#kcs">Similarity to Known Clusters</a>
            </li>
            <li>
                <a data-toggle="pill" href="#sh">BGC Self-Similarity</a>
            </li>
            <li>
                <a data-toggle="pill" href="#pd">Phylogenetic Diversity</a>
            </li>
            <li>
                <a data-toggle="pill" href="#ts">Technical Specifications</a>
            </li>
        </ul>
    </div>

    <div class="tab-content" style="display: table-cell; padding-left: 30px; overflow: scroll;">
        <div id="helpHome" class="tab-pane fade in active" style="width: 850px">
            <h3>About antiSMASH - OOPS</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    antiSMASH - OOPS provides a reliable, fully functional, structured and flexible web application for
                    the filtering and ranking of Biosynthetic Gene Clusters according by a wide variety of custom-weighted
                    biological and biochemical criteria.
                    <br/><br/>
                    This computational tool is a Biosynthetic Gene Cluster prioritisation system that takes for input one
                    or more compressed (zip) files from antiSMASH and processes them according to several user defined
                    parameters such as number of genes, CDS Length, GC Content, Codon Bias, Cluster Type, Similarity to
                    Known Clusters, Self-Similarity and Phylogenetic Diversity. For each parameter, a detailed description
                    is provided in the following sections.
                    <br/><br/>
                    This project was developed by Alejandro Pena as Dissertation Project for the MSc degree in Advanced
                    Computer Science with close mentoring and supervision of Dr. Rainer Breitling, professor of Systems
                    Biology in the University of Manchester and researcher at the Manchester Institute of Biotechnology.
                </p>
            </div>
        </div>
        <div id="nog" class="tab-pane fade" style="width: 850px">
            <h3>Number of Genes</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    The prioritisation according to this parameter is straightforward. By default clusters with the highest
                    gene number are ranked first. The order can be reversed by clicking in the small prioritisation order icon.
                    The weight of this criterion in the prioritisation can be defined by the user using the corresponding slider.
                </p>
            </div>
        </div>
        <div id="sl" class="tab-pane fade" style="width: 850px">
            <h3>CDS Length</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    This parameter allows the user to rank the clusters according to the length of their total coding DNA sequence.
                    Again, the user can change both the ranking order and the weight of this criterion.
                </p>
            </div>
        </div>
        <div id="gcc" class="tab-pane fade" style="width: 850px">
            <h3>GC Content</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    The G+C Content parameter can be used both alone and in conjunction with the Reference Species input
                    parameter. If no reference species is provided, the application will calculate the GC Content of the
                    Biosynthetic Gene Cluster and use these values to rank the clusters. Whereas, if a reference species
                    is provided, the application will retrieve the GC Content of that species and the cluster are ranked
                    according to the difference of GC content between clusters as reference specie. The order icon and
                    the weight slider can be used to choose the prioritization order and the weight of this parameter on
                    the prioritisation.
                </p>
            </div>
        </div>
        <div id="cb" class="tab-pane fade" style="width: 850px">
            <h3>Codon Bias</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    This parameter must be used in conjunction with the Reference Species input parameter. If no species
                    is specified, the prioritisation of the Biosynthetic Gene Clusters according to this parameter will
                    be ignored and the only output will be the Codon Usage Table of the respective BGC. If a reference
                    species is provided, the application will build a Codon Bias Heat Chart indicating in blue the codon
                    usage bias towards the BGC and in green the bias towards the reference species. Also for this parameter,
                    it is possible to set the prioritisation order and its weight in the prioritisation.
                </p>
            </div>
        </div>
        <div id="rs" class="tab-pane fade" style="width: 850px">
            <h3>Reference Species</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    The application allows selecting a reference species for the prioritisation, which affects two parameters:
                    the GC content and the Codon Bias.
                </p>
            </div>
        </div>
        <div id="ct" class="tab-pane fade" style="width: 850px">
            <h3>Cluster Type</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    This parameter will filter the Biosynthetic Gene Clusters according to their type. Only the selected
                    cluster types in the Select Box will be considered in the prioritisation. It is possible to select
                    one, multiple or none types in the Select Box. If no types are selected it will be equivalent to tick
                    the Ignore combo box for this parameter and all the BGCs loaded to the application will be processed.
                </p>
            </div>
        </div>
        <div id="kcs" class="tab-pane fade" style="width: 850px">
            <h3>Similarity to Known Clusters</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    antiSMASH - OOPS takes the information about the similar Known Clusters found for each Biosynthetic
                    Gene Cluster form the antiSMASH output and computes a similarity score considering gene identity,
                    gene coverage and number of genes found in the known cluster. The clusters will be ranked according
                    to the difference between the selected preferred similarity and the highest similarity to a known
                    cluster found. It is also possible to define a +/- parameter which defines a range around the preferred
                    similarity. All values included in such range will be consider equal for the purposes of the ranking.
                    It is possible as well to set the prioritisation process to be ascending or descending order. Again,
                    the user can change both the ranking order and the weight of this criterion.
                </p>
            </div>
        </div>
        <div id="sh" class="tab-pane fade" style="width: 850px">
            <h3>BGC Self-Similarity</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    This parameter calculates a score based on how similar a Biosynthetic Gene Cluster is to itself.
                    It uses a modified version of the Smith-Waterman algorithm to find local alignments between two
                    sequences. The clusters are ranked according to the sum of the length of all suboptimal alignments
                    found. It is possible to specify a minimum alignment number of nucleotides as input preference for
                    the Self-Similarity parameter, if selected all alignment shorter than that will be ignored. In the
                    output, OOPS shows also the length of the longest alignment. Weight and order can be selected by
                    the user.
                </p>
            </div>
        </div>
        <div id="pd" class="tab-pane fade" style="width: 850px">
            <h3>Phylogenetic Diversity</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    This parameter takes a Biosynthetic Gene Cluster, gets the one hundred top cluster BLAST matches
                    done by antiSMASH and builds a single tree of life with the lineage of each one of the clusters.
                    This operation is repeated for every single BGC uploaded to antiSMASH - OOPS and once all the trees
                    are built, OOPS computes the phylogenetic diversity score as the number of nodes of the tree above
                    the species level. Again, the user can change both the ranking order and the weight of this criterion.
                </p>
            </div>
        </div>
        <div id="ts" class="tab-pane fade" style="width: 850px">
            <h3>Technical Specifications</h3>
            <div style="background-color: #E0E0E0; padding: 10px">
                <p>
                    antiSMASH - OOPS is a Java based Web Application developed using as main engine the Spring 4 Framework
                    for handling the business logic of the system. The front end is developed using HTML, JavaScript, JQuery,
                    AJAX, Bootstrap, the HighCharts Framework and Thymeleaf as the Java templating engine.
                    <br/><br/>
                    This tool is packaged in a single JAR file which contains an embedded Tomcat Web Server that makes it
                    possible to run literally in any environment that supports a Java Virtual Machine. Even though the
                    application is distributed in a JAR file, the Spring Boot Framework allows it to become a web application
                    easily and without the hassle of configuring a web server and deploying the project in it.
                    <br/><br/>
                    The JAR file includes all the necessary libraries and frameworks that the application needs to run properly
                    and the server is started just by double clicking in the JAR file on Windows. Alternatively to launch the
                    OOPS application, open either the terminal (on Linux/MacOS) or the command prompt (on Windows: Start >
                    All Programmes > Accessories > Command Prompt) and navigate to the folder containing the antiSMASH-OOPS.jar
                    file, then type the following:
                    <br/><br/>
                    <i>java -jar antiSMASH-OOPS.jar</i>
                    <br/><br/>
                    and hit enter. The application will load. Once the application has loaded one can view the user interface
                    by opening a web browser (Chrome or Safari recommended) and typing the following in the URL bar before
                    pressing enter:
                    <br/><br/>
                    <b>localhost:8080</b>
                    <br/><br/>
                    The landing page of the system will show up immediately and will list any available workspace.
                </p>
            </div>
        </div>
    </div>

</div>


<div th:replace="fragments/footer :: footer"></div>
</body>
</html>