<!DOCTYPE html>
<html xmlns:th="http://www.thymeleaf.org">
<head>
<meta charset="UTF-8"></meta>
<title th:inline="text">[[#{app.name}]] :: Help</title>

<link href="http://cdn.jsdelivr.net/webjars/bootstrap/3.3.6/css/bootstrap.min.css"
	th:href="@{/webjars/bootstrap/3.3.6/css/bootstrap.min.css}" rel="stylesheet" media="screen"
/>
<link rel="stylesheet" th:href="@{/css/application.css}" />
<script src="http://cdn.jsdelivr.net/webjars/jquery/2.1.3/jquery.min.js" th:src="@{/webjars/jquery/2.1.3/jquery.min.js}"></script>
<script th:src="@{/js/script.js}"></script>
<script th:src="@{/js/bootstrap.min.js}"></script>

<link rel="icon" th:href="@{/images/oops.ico}" />

</head>

<body>
	<div th:replace="fragments/header :: header"></div>

	<div class="container">
		<div class="div-ctr">
			<h2 th:inline="text">HELP PAGE</h2>
		</div>
		<br />

		<div style="display: table-cell; padding-right: 30px; border-right: solid; border-color: lightgrey">
			<ul class="nav nav-pills nav-stacked">
				<li class="active">
					<a data-toggle="pill" href="#helpHome">antiSMASH - OOPS</a>
				</li>
				<li>
					<a data-toggle="pill" href="#nog">Number of Genes</a>
				</li>
				<li>
					<a data-toggle="pill" href="#sl">CDS Length</a>
				</li>
				<li>
					<a data-toggle="pill" href="#gcc">GC Content</a>
				</li>
				<li>
					<a data-toggle="pill" href="#cb">Codon Bias</a>
				</li>
				<li>
					<a data-toggle="pill" href="#rs">Reference Species</a>
				</li>
				<li>
					<a data-toggle="pill" href="#ct">Cluster Type</a>
				</li>
				<li>
					<a data-toggle="pill" href="#kcs">Similarity to Known Clusters</a>
				</li>
				<li>
					<a data-toggle="pill" href="#sh">BGC Self-Homology</a>
				</li>
				<li>
					<a data-toggle="pill" href="#pd">Phylogenetic Diversity</a>
				</li>
				<li>
					<a data-toggle="pill" href="#ts">Technical Specifications</a>
				</li>
			</ul>
		</div>

		<div class="tab-content" style="display: table-cell; padding-left: 30px; overflow: scroll;">
			<div id="helpHome" class="tab-pane fade in active" style="width: 850px">
				<h3>About antiSMASH - OOPS</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					The aim of antiSMASH - OOPS is to is to provide for biologists a reliable, fully functional and
					structured computational tool in the form of a web application, that allows them to filter and 
					extract the best Biosynthetic Gene Clusters according to their probablity of producing a novel
				    end product of interest.
				    <br />
					<br />
					This computational tool is a Biosynthetic Gene Cluster prioritisation system that takes for input
					one or more compressed (zip) files from antiSMASH and processes them according to several user defined
					parameters such as number of genes, CDS Length, GC Content, Codon Bias, Cluster Type, Similarity
					to Known Clusters, Self-Homology and Phylogenetic Diversity.
					<br />
					<br />
					Each one of those parameters is explained in detail in the following sections as well as information
					on how are the clusters scored based on those.
					<br />
					<br />
					This project was developed by Alejandro Pena as Dissertation Project for the MSc degree in Advanced
					Computer Science with close mentoring and supervision of Dr. Rainer Breitling, professor of Systems
					Biology in the University of Manchester and researcher at the Manchester Institute of Biotechnology. 
				</p>
				</div>
			</div>
			<div id="nog" class="tab-pane fade" style="width: 850px">
				<h3>Number of Genes</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					Prioritisation according to this parameter is fairly straightforward. The clusters with the biggest
					number of genes detected are ranked higher than those with fewer number of genes. This ranking can
					be reversed by clicking in the small prioritisation order icon. It is possible to set the proritisation
					process to be ascending or descending order. You should move the corresponding slider to tell the
					application how much are you interested in this parameter and the system will make the appropriate
					ponderations.
				</p>
				</div>
			</div>
			<div id="sl" class="tab-pane fade" style="width: 850px">
				<h3>CDS Length</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					The CDS Length parameter scores the Biosynthetic Gene Clusters by getting all the coding sequences
					within the cluster, adding them up and setting by default as top most cluster the one with the longest
					sequence. It is possible to set the proritisation process to be ascending or descending
					order. You should move the corresponding slider to tell the application how much are you interested
					in this parameter and the system will make the appropriate ponderations.
				</p>
				</div>
			</div>
			<div id="gcc" class="tab-pane fade" style="width: 850px">
				<h3>GC Content</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					The GC Content parameter can be used in conjunction with the Reference Species input parameter. If no
					reference species is provided the application will calculate the GC Content of the Biosynthetic Gene
					Cluster and show as top most result the one that has a higher content whereas with lower content the
					score will be lower as well. If a reference species is provided, the application will retrieve the
					GC Content of that species and will score a BGC higher if the difference with regard to the reference
					species GC Content is low. On the contrary, if the difference against the reference species is
					high the BGC will be scored lower. It is possible to set the proritisation process to be ascending
					or descending order. You should move the corresponding slider to tell the application how much are you
					interested in this parameter and the system will make the appropriate ponderations.
				</p>
				</div>
			</div>
			<div id="cb" class="tab-pane fade" style="width: 850px">
				<h3>Codon Bias</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					This parameter must be used in conjunction with the Reference Species input parameter. If no species
					is specified, the prioritisation of the Biosynthetic Gene Clusters according to this parameter will be
					ignored and the output will be only the Codon Usage Table of the respective BGC. If a reference species
					is provided, the application will build a Codon Bias Heat Chart indicating in blue color the codon usage
					bias towards the BGC and in green the bias towards the reference species. For this parameter, a BGC that
					results to be less biased towards any of the two sides will be scored higher, and a BGC with a high
					bias will be scored lower. It is possible to set the proritisation process to be ascending or descending
					order. You should move the corresponding slider to tell the application how much are you interested
					in this parameter and the system will make the appropriate ponderations.
				</p>
				</div>
			</div>
			<div id="rs" class="tab-pane fade" style="width: 850px">
				<h3>Reference Species</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					The application allows to select a reference species for the prioritisation. This reference species
					affects two parameters. The GC Content and the Codon Bias parameter. If no reference species is
					selected the Codon Bias parameter will be ignored and the prioritisation according to the GC content
					parameter will be done considering only the Biosynthetic Gene Cluster GC Content without making any
					comparison against a reference species.
				</p>
				</div>
			</div>
			<div id="ct" class="tab-pane fade" style="width: 850px">
				<h3>Cluster Type</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					This parameter will filter the Biosynthetic Gene Clusters according to their type. Only the selected
					clusters in the Select Box will show as output for the prioritisation. It is possible to select one,
					multiple or none types in the Select Box. If no types are selected it will be equivalent to tick the
					Ignore combo box for this parameter and all the BGCs loaded to the application will be processed.
				</p>
				</div>
			</div>
			<div id="kcs" class="tab-pane fade" style="width: 850px">
				<h3>Similarity to Known Clusters</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					antiSMASH - OOPS takes data of the Known Clusters found for each Biosynthetic Gene Cluster outputted by
					antiSMASH and generates a score of similarity considering gene identity, gene coverage and number of
					genes found in the known cluster. It is possible to specify a preferred percentage of similarity to
					the known clusters. It is possible as well to set the proritisation process to be ascending or descending
					order. You should move the corresponding slider to tell the application how much are you interested
					in this parameter and the system will make the appropriate ponderations.
				</p>
				</div>
			</div>
			<div id="sh" class="tab-pane fade" style="width: 850px">
				<h3>BGC Self-Homology</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					This parameter calculates a score based on how similar a Biosynthetic Gene Cluster is to itself. It
					uses a modified version of the Smith-Waterman algorithm to find local alignments between two sequences
					however for this case, since the sequence is aligned to itself, the suboptimal alignments are the ones.
					that the algorithm finds. It is possible to specify a minimum alignment number of nucleotides as input 
					preference for the Self-Homology parameter. It is possible as well  to set the proritisation process to be
					ascending or descending	order. You should move the corresponding slider to tell the application how
					much are you interested	in this parameter and the system will make the appropriate ponderations.
				</p>
				</div>
			</div>
			<div id="pd" class="tab-pane fade" style="width: 850px">
				<h3>Phylogenetic Diversity</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					This parameter takes a Biosynthetic Gene Cluster, gets the one hundred top cluster BLAST matches done by
					antiSMASH and builds a single tree of life with the lineage of each one of the clusters. This operation
					is repeated for every single BGC uploaded to antiSMASH - OOPS and once all the trees are built, it
					calculates the size of the tree formed. Having a bigger score means that the Phylogenetic Diversity is
					higher and getting a smaller score indicates that the BGC blasts are less diverse phylogenetically.
					It is possible to set the proritisation process to be ascending or descending order. You should move
					the corresponding slider to tell the application how much are you interested in this parameter and the
					system will make the appropriate ponderations.
				</p>
				</div>
			</div>
			<div id="ts" class="tab-pane fade" style="width: 850px">
				<h3>Technical Specifications</h3>
				<div style="background-color: #E0E0E0; padding: 10px">
				<p>
					antiSMASH - OOPS is a Java based Web Application developed using as main engine the Spring 4 Framework
					for handling the business logic of the system. The front end is developed using HTML, JavaScript, JQuery,
					AJAX, Bootstrap, the HighCharts Framework and Thymeleaf as the Java templating engine.
					<br />
					<br />
					This tool is packaged in a single JAR file which contains an embedded Tomcat Web Server that makes it
					possible to run literally in any environment that supports a Java Virtual Machine. Even though the 
					application is distributed in a JAR file, the Spring Boot Framework allows it to become a web application
					easily and without the hassle of configuring a web server and deploying the project in it.
					<br />
					<br />
					The JAR file includes all the necessary libraries and frameworks that the application needs to run properly
					and the server is started just by double clicking in the JAR file. It is possible to start the application
					by running the StartApplication script and stop the application in the same way by running the StopApplication
					script. To access antiSMASH - OOPS just type in any browser -localhost:8080- and the main page will be loaded
				</p>
				</div>
			</div>
		</div>

	</div>



	<div th:replace="fragments/footer :: footer"></div>
</body>
</html>